lib/Builtins.cpp

```diff
@@ -300,6 +307,101 @@ std::string Builtins::GetMangledFunctionName(const char *name) {
   return std::string("_Z") + std::to_string(strlen(name)) + name;
 }
 
+std::string
+Builtins::GetMangledFunctionName(const Builtins::FunctionInfo &info) {
+  // This is a best-effort attempt at reconstructing the mangled name for the
+  // given function. Because demangling is a lossy process some information may
+  // be lost and is therefore no longer available.
+  std::string name;
+  raw_string_ostream out(name);
+
+  auto print_name = [&out](StringRef str) {
+    out << "_Z" << std::to_string(str.size()) << str;
+  };
+
+  print_name(info.getName());
+
+  for (size_t i = 0; i < info.getParameterCount(); ++i) {
+    const auto &param = info.getParameter(i);
+
+    if (param.vector_size != 0) {
+      out << "Dv" << std::to_string(param.vector_size);
+    }
+
+    switch (param.type_id) {
+    case Type::FloatTyID:
+      switch (param.byte_len) {
+      case 2:
+        out << "Dh";
+        break;
+      case 4:
+        out << "f";
+        break;
+      case 8:
+        out << "d";
+        break;
+      default:
+        llvm_unreachable("Invalid byte_len for floating point type.");
+        break;
+      }
+      break;
+
+    case Type::IntegerTyID:
+      if (param.is_signed) {
+        switch (param.byte_len) {
+        case 1:
+          // Not enough information to distinguish between char (c) and signed
+          // char (a).
+          out << 'c';
+          break;
+        case 2:
+          out << "s";
+          break;
+        case 4:
+          out << "i";
+          break;
+        case 8:
+          out << "l";
+          break;
+        default:
+          llvm_unreachable("Invalid byte_len for signed integer type.");
+          break;
+        }
+      } else {
+        switch (param.byte_len) {
+        case 1:
+          out << 'h';
+          break;
+        case 2:
+          out << "t";
+          break;
+        case 4:
+          out << "j";
+          break;
+        case 8:
+          out << "m";
+          break;
+        default:
+          llvm_unreachable("Invalid byte_len for unsigned integer type.");
+          break;
+        }
+      }
+      break;
+
+    case Type::StructTyID:
+      print_name(param.name);
```

Comment:
You were right. I've fixed this, and another bug (the return statement was within the for loop, not after it). Note that this `case` isn't exercised with the existing tests -- I'm not sure there is an OpenCL function with both a struct and a long-vector as parameters and without such builtin, the lowering phase will not trigger this `case`.

It might be good to have some light unit testing for this functionality, but I'm not sure where to start. I've tested this manually for now. I suppose this can be improved upon later as a general issue if desired.